tonic/src/client/grpc.rs

```diff
@@ -28,12 +30,102 @@ use std::fmt;
 /// [gRPC protocol definition]: https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md#requests
 pub struct Grpc<T> {
     inner: T,
+    #[cfg(feature = "compression")]
+    /// Which compression encodings does the client accept?
+    accept_compression_encodings: EnabledCompressionEncodings,
+    #[cfg(feature = "compression")]
+    /// The compression encoding that will be applied to requests.
+    send_compression_encodings: Option<CompressionEncoding>,
 }
 
 impl<T> Grpc<T> {
     /// Creates a new gRPC client with the provided [`GrpcService`].
     pub fn new(inner: T) -> Self {
-        Self { inner }
+        Self {
+            inner,
+            #[cfg(feature = "compression")]
+            send_compression_encodings: None,
+            #[cfg(feature = "compression")]
+            accept_compression_encodings: EnabledCompressionEncodings::default(),
+        }
+    }
+
+    /// Compress requests with `gzip`.
+    ///
+    /// Requires the server to accept `gzip` otherwise it might return an error.
+    ///
+    /// # Example
+    ///
+    /// The most common way of using this is through a client generated by tonic-build:
+    ///
+    /// ```rust
+    /// use tonic::transport::Channel;
+    /// # struct TestClient<T>(T);
+    /// # impl<T> TestClient<T> {
+    /// #     fn new(channel: T) -> Self { Self(channel) }
+    /// #     fn send_gzip(self) -> Self { self }
+    /// # }
+    ///
+    /// # async {
+    /// let channel = Channel::builder("127.0.0.1:3000".parse().unwrap())
+    ///     .connect()
+    ///     .await
+    ///     .unwrap();
+    ///
+    /// let client = TestClient::new(channel).send_gzip();
+    /// # };
+    /// ```
+    #[cfg(feature = "compression")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "compression")))]
+    pub fn send_gzip(mut self) -> Self {
+        self.send_compression_encodings = Some(CompressionEncoding::Gzip);
+        self
+    }
+
+    #[doc(hidden)]
+    #[cfg(not(feature = "compression"))]
+    pub fn send_gzip(self) -> Self {
+        panic!(
+            "`send_gzip` called on a server but the `compression` feature is not enabled on tonic"
```

Comment:
This mentions server but we are in the client? Also, wouldn't it be better to just have this not compile, aka just don't include a `not` version.

Reply:
Yes not compiling would be better, and its also what I did initially, but it doesn't work with the ["Check features"](https://github.com/hyperium/tonic/blob/master/.github/workflows/CI.yml#L32) step in CI. It seems to test all combinations of features. It doesn't seem to be possible to tell it to exclude a set of features.